Lib/pty.py

```diff
@@ -170,3 +175,128 @@ def spawn(argv, master_read=_read, stdin_read=_read):
 
     os.close(master_fd)
     return os.waitpid(pid, 0)[1]
+
+# Author: Soumendra Ganguly.
+SIGWINCH = signal.SIGWINCH
+
+def _login_pty(master_fd, slave_fd):
+    """Given a pty, makes the calling process a session leader, makes the pty slave
+    its controlling terminal, stdin, stdout, and stderr. Closes both pty ends."""
+    # Establish a new session.
+    os.setsid()
+    os.close(master_fd)
+
+    try:
+        fcntl.ioctl(slave_fd, termios.TIOCSCTTY) # Make the pty slave the controlling terminal.
+    except:
+        os.close(slave_fd)
+        raise
+
+    # Slave becomes stdin/stdout/stderr.
+    os.dup2(slave_fd, STDIN_FILENO)
+    os.dup2(slave_fd, STDOUT_FILENO)
+    os.dup2(slave_fd, STDERR_FILENO)
+    if (slave_fd > STDERR_FILENO):
+        os.close(slave_fd)
+
+def _winresz(pty_slave):
+    """Resize window."""
+    w = struct.pack('HHHH', 0, 0, 0, 0)
+    s = fcntl.ioctl(STDIN_FILENO, termios.TIOCGWINSZ, w)
+    fcntl.ioctl(pty_slave, termios.TIOCSWINSZ, s)
+
```

Comment:
The SIGWINCH handler is the star of the show here. The handler depends on the pty slave returned by openpty in the body of "wspawn"; while the handler could just have been a function nested inside wspawn, that would make wspawn less readable. That is the reason for writing "_create_hwinch", which simply takes the pty slave as an argument and returns the appropriate signal handler function [ for SIGWINCH ].